subscription.go

```diff
@@ -163,39 +163,80 @@ func (r *marathonClient) registerCallbackSubscription() error {
 }
 
 func (r *marathonClient) registerSSESubscription() error {
-	// Prevent multiple SSE subscriptions
 	if r.subscribedToSSE {
 		return nil
 	}
 
-	request, _, err := r.buildAPIRequest("GET", marathonAPIEventStream, nil)
+	stream, err := r.connectToSSE()
 	if err != nil {
 		return err
 	}
+	go r.listenToSSE(stream)
 
-	// Try to connect to stream, reusing the http client settings
-	stream, err := eventsource.SubscribeWith("", r.config.HTTPClient, request)
-	if err != nil {
-		return err
-	}
+	r.subscribedToSSE = true
+	return nil
+}
 
-	go func() {
-		for {
-			select {
-			case ev := <-stream.Events:
-				if err := r.handleEvent(ev.Data()); err != nil {
-					// TODO let the user handle this error instead of logging it here
-					r.debugLog.Printf("registerSSESubscription(): failed to handle event: %v\n", err)
-				}
-			case err := <-stream.Errors:
+func (r *marathonClient) listenToSSE(stream *eventsource.Stream) {
+	for {
+		select {
+		case ev := <-stream.Events:
+			if err := r.handleEvent(ev.Data()); err != nil {
 				// TODO let the user handle this error instead of logging it here
-				r.debugLog.Printf("registerSSESubscription(): failed to receive event: %v\n", err)
+				r.debugLog.Printf("listenToSSE(): failed to handle event: %v\n", err)
 			}
+		case err := <-stream.Errors:
+			r.debugLog.Printf("listenToSSE(): received error from SSE stream: %s", err)
+			stream.Close()
+			stream = r.mustConnectToSSE()
 		}
-	}()
+	}
+}
 
-	r.subscribedToSSE = true
-	return nil
+// connectToSSE tries to establish an *eventsource.Stream to any of the Marathon cluster members, marking the
+// member as down on connection failure, until there is no more active member in the cluster.
+func (r *marathonClient) connectToSSE() (*eventsource.Stream, error) {
+	for {
+		request, member, err := r.buildAPIRequest("GET", marathonAPIEventStream, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		// The event source library manipulates the HTTPClient. So we create a new one and copy
+		// its underlying fields for performance reasons. See note that at least the Transport
+		// should be reused here: https://golang.org/pkg/net/http/#Client
+		httpClient := &http.Client{
+			Transport:     r.config.HTTPClient.Transport,
+			CheckRedirect: r.config.HTTPClient.CheckRedirect,
+			Jar:           r.config.HTTPClient.Jar,
+			Timeout:       r.config.HTTPClient.Timeout,
+		}
+
+		stream, err := eventsource.SubscribeWith("", httpClient, request)
+		if err != nil {
+			r.hosts.markDown(member)
+			continue
+		}
+
+		return stream, nil
+	}
+}
+
+// mustConnectToSSE will attempt to connect to the SSE stream until it can be established.
+// It panics when there is an error building the HTTP request.
+func (r *marathonClient) mustConnectToSSE() *eventsource.Stream {
+	for {
+		if len(r.hosts.activeMembers()) > 0 {
+			stream, err := r.connectToSSE()
+			if err == nil {
+				return stream
+			}
+			if err != ErrMarathonDown {
```

Comment:
Two thoughts on this one:

1. Should we have this check at all? I'm worried about a hypothetical future case where `buildAPIRequest` (which is called by `connectToSSE`) is extended to return a non-nil error value other than `ErrMarathonDown`.
1. If we think the invariant still holds: Should it maybe rather happen in `connectToSSE`?